### PR TITLE
parsing plugins remodularization

### PR DESCRIPTION
**Pull request by AgustinBettati**
This PR includes:
- APIMF-2877: definition of parsing plugins and adoption in AMFCompiler
- APIMF-2905: remove validation methods from parse plugin interface
- APIMF-2907: create default environments with parsing plugins
- APIMF-2880: define interface for fallback mechanism when no plugin is matched

depends on https://github.com/aml-org/amf-aml/pull/198 and https://github.com/aml-org/amf-core/pull/178

### CODE REVIEW

File: amf-client/shared/src/main/scala/amf/client/commands/CommandHelper.scala [hunk 1]
```diff
@@ -28,6 +39,7 @@ trait CommandHelper {
     amf.core.AMF.registerPlugin(Raml08Plugin)
     amf.core.AMF.registerPlugin(Oas20Plugin)
     amf.core.AMF.registerPlugin(Oas30Plugin)
```
Comment (tomsfernandez) [inline amf-client/shared/src/main/scala/amf/client/commands/CommandHelper.scala:29]: [OOS] - Not part of REMOD, could be taken to another commit.

File: amf-client/shared/src/test/resources/validations/reports/model/invalid-reference-defined-in-lib.report [hunk 2]
```diff
@@ -0,0 +1,38 @@
+Model: file://amf-client/shared/src/test/resources/validations/invalid-reference/error-in-lib.raml
+Profile: RAML 1.0
+Conforms? false
+Number of results: 4
```
Comment (tomsfernandez) [inline amf-client/shared/src/test/resources/validations/reports/model/invalid-reference-defined-in-lib.report:4]: Why is there a new test?

File: amf-client/shared/src/test/scala/amf/cycle/OasRecursiveFilesCycleTest.scala [hunk 3]
```diff
@@ -43,7 +42,7 @@ class OasRecursiveFilesCycleTest extends FunSuiteCycleTests {
 
       RuntimeCompiler
         .forContext(
-          compilerContextBuilder.withEnvironment(Environment.apply()).build(),
+          compilerContextBuilder.build(),
```
Comment (tomsfernandez) [inline amf-client/shared/src/test/scala/amf/cycle/OasRecursiveFilesCycleTest.scala:45]: Can we break this interface?

File: amf-client/shared/src/test/scala/amf/error/RamlParserErrorTest.scala [hunk 4]
```diff
@@ -98,19 +98,15 @@ class RamlParserErrorTest extends ParserErrorTest {
   test("Inline external fragment from non mutable ref") {
     validate(
       "error/inline-non-mutable-ref/api.raml",
-      badInclude1 => {
-        badInclude1.level should be("Violation")
-        badInclude1.message should startWith("Fragments must be imported by using '!include'")
-      },
-      badInclude2 => {
```
Comment (tomsfernandez) [inline amf-client/shared/src/test/scala/amf/error/RamlParserErrorTest.scala:105]: Why do we have one less assertion?

File: amf-webapi/shared/src/main/scala/amf/client/environment/ApiEnvironment.scala [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package amf.client.environment
+
+import amf.client.remod.AMFEnvironment
+import amf.plugins.document.graph.AMFGraphParsePlugin
+import amf.plugins.document.webapi._
+
+// defined private unit mayor version is released
+private[amf] object ApiEnvironment {
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/client/environment/ApiEnvironment.scala:8]: Hmm suspicious. With this we are forcing adopters to depend on all of the artifacts we support. I agree that we have to make it easy for adopters to migrate but still, suspicious.

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/AsyncPlugin.scala [hunk 6]
```diff
@@ -35,10 +35,7 @@ sealed trait AsyncPlugin extends OasLikePlugin {
               wrapped: ParserContext,
               ds: Option[AsyncWebApiDeclarations] = None): AsyncWebApiContext
 
-  override def parse(document: Root,
-                     parentContext: ParserContext,
-                     platform: Platform,
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/AsyncPlugin.scala:40]: I saw Platform being used as a parameter in several places, why do we have this if Platform is a Singleton for the distribution? Is this usage legacy?

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/references/RamlReferenceHandler.scala [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  /** Update parsed reference if needed. */
+  override def update(reference: ParsedReference, compilerContext: CompilerContext)(
+      implicit executionContext: ExecutionContext): Future[ParsedReference] = {
+    if (reference.isExternalFragment)
+      handleRamlExternalFragment(reference, compilerContext)
+    else Future.successful(reference)
+  }
+
+  private def handleRamlExternalFragment(reference: ParsedReference, compilerContext: CompilerContext)(
+      implicit executionContext: ExecutionContext): Future[ParsedReference] = {
```
Comment (tomsfernandez) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/references/RamlReferenceHandler.scala:31]: [OOS] - A nice cleanup could be done here.

**Review by tomsfernandez — DISMISSED** [review]

File: amf-client/shared/src/test/resources/validations/reports/model/invalid-reference-defined-in-lib.report (hunk 2 above)
Comment (AgustinBettati) [reply amf-client/shared/src/test/resources/validations/reports/model/invalid-reference-defined-in-lib.report:4]: This was included when moving verifyValidReferenceKind from the document plugin interface to the raml plugin, i found a specific case that was not being tested so i included this test case (using invalid references within in a library).

File: amf-client/shared/src/test/scala/amf/cycle/OasRecursiveFilesCycleTest.scala (hunk 3 above)
Comment (AgustinBettati) [reply amf-client/shared/src/test/scala/amf/cycle/OasRecursiveFilesCycleTest.scala:45]: This interface has not been broken (CompilerContextBuilder), removed this code as it was useless.

File: amf-client/shared/src/test/scala/amf/error/RamlParserErrorTest.scala (hunk 4 above)
Comment (AgustinBettati) [reply amf-client/shared/src/test/scala/amf/error/RamlParserErrorTest.scala:105]: This was due to the change of verifyValidReferenceKind. This test case has a an external fragment that references a raml library incorrectly. This incorrect reference used to be validated with the raml plugin (even though the file in not raml related), while now as these reference validation are specific for raml it is no longer present.
We analyzed this behavior change with @pope1838 and agreed that the impact is very reduced and can be considered correct.

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/AsyncPlugin.scala (hunk 6 above)
Comment (AgustinBettati) [reply amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/AsyncPlugin.scala:40]: Introduced this changed as it was a parameter that was unused in all plugin implementations. As you mention, this most likely was a legacy usage.

File: amf-webapi/shared/src/main/scala/amf/client/parse/Oas20Parser.scala [hunk 8]
```diff
@@ -3,7 +3,7 @@ package amf.client.parse
 import amf.client.environment.Environment
 import amf.core.registries.AMFPluginsRegistry
 import amf.core.remote.Oas20
-import amf.plugins.document.webapi.Oas20Plugin
+import amf.plugins.document.webapi.{Async20ParsePlugin, ExternalJsonYamlRefsParsePlugin, Oas20ParsePlugin, Oas20Plugin}
```
Comment (looseale) [inline amf-webapi/shared/src/main/scala/amf/client/parse/Oas20Parser.scala]: Delete import of Async20ParsePlugin, also ExternalJsonYamlRefsParsePlugin.
In general all parsers looks to have imports of plugins that are not used (were added in a previous commit but the uses were deleted in a newer one)

**Review by looseale — DISMISSED** [review]

**Review by looseale — APPROVED** [review]
